### PR TITLE
Tuples

### PR DESCRIPTION
**Pull request by dabrahams**

### CODE REVIEW

File: Sources/PenguinStructures/Tuple.swift [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  var head: Head { get set }
+  
+  /// All elements but the first.
+  var tail: Tail { get set }
+}
+
+extension Empty: AlgebraicProduct {
+  /// The first element, when `self` is viewed as an instance of algebraic
+  /// product type.
+  public var head: Never { get {  fatalError() } set {  } }
```
Comment (saeta) [inline Sources/PenguinStructures/Tuple.swift:39]: Should `set` also `fatalError()`?

File: Sources/PenguinStructures/Tuple.swift [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+public typealias Tuple6<T0, T1, T2, T3, T4, T5>
+  = Tuple<T0, Tuple5<T1, T2, T3, T4, T5>>
+public typealias Tuple7<T0, T1, T2, T3, T4, T5, T6>
+  = Tuple<T0, Tuple6<T1, T2, T3, T4, T5, T6>>
+
+public extension Tuple {
+  /// Creates an instance containing the arguments, in order.
+  init(_ a0: Head) where Tail == Empty {
+    head = a0; tail = Tuple0()
+  }
```
Comment (saeta) [inline Sources/PenguinStructures/Tuple.swift:107]: I think you can get this to compile by writing:

```suggestion
public extension Tuple where Tail == Empty {
  /// Creates an instance containing the arguments, in order.
  init(_ a0: Head) {
    head = a0; tail = Tuple0()
  }
}
public extension Tuple {
```

File: Sources/PenguinStructures/Tuple.swift (hunk 1 above)
Comment (dabrahams) [reply Sources/PenguinStructures/Tuple.swift:39]: Neither set nor get can be called; the `fatalError` is just there to make the compiler happy.

File: Sources/PenguinStructures/Tuple.swift (hunk 2 above)
Comment (dabrahams) [reply Sources/PenguinStructures/Tuple.swift:107]: Could swear I did compile it.  Heh, works with our Swift but not Apple's.  Thanks!

File: Tests/PenguinStructuresTests/TupleTests.swift [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      type(of: Tuple(x0, x1, x2, x3, x4, x5))
+        == Tuple6<X0, X1, X2, X3, X4, X5>.self)
+    XCTAssert(
+      type(of: Tuple(x0, x1, x2, x3, x4, x5, x6))
+        == Tuple7<X0, X1, X2, X3, X4, X5, X6>.self)
+  }
+  
+  static var allTests = [
+    ("test_mapReduce", test_mapReduce),
+  ]
```
Comment (saeta) [inline Tests/PenguinStructuresTests/TupleTests.swift:165]:
```suggestion
  static var allTests = [
    ("test_mapReduce", test_mapReduce),
    ("test_head", test_head),
    ("test_tail", test_tail),
    ("test_count", test_count),
    ("test_DefaultInitializable", test_DefaultInitializable),
    ("test_Equatable", test_Equatable),
    ("test_Comparable", test_Comparable),
    ("test_CustomStringConvertible", test_CustomStringConvertible),
    ("test_conveniences", test_conveniences),
  ]
```

File: Sources/PenguinStructures/Tuple.swift [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    else {
+      return "Tuple(\(String(reflecting: head)), "
+        + String(reflecting: tail).dropFirst(prefixLength)
+    }
+  }
+}
+
+// ======== Conveniences ============
+
+public typealias Tuple0 = Empty
```
Comment (saeta) [inline Sources/PenguinStructures/Tuple.swift:104]: Might we want to add doc comments to these in some form?

File: Tests/PenguinStructuresTests/TupleTests.swift [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    
+    XCTAssertEqual(Tuple(1, 2), Tuple(1, 2))
+    XCTAssertNotEqual(Tuple(0, 2), Tuple(1, 2))
+    XCTAssertNotEqual(Tuple(1, "2"), Tuple(1, "XXX"))
+    
+    XCTAssertEqual(Tuple(1, 2.3, "4"), Tuple(1, 2.3, "4"))
+    XCTAssertNotEqual(Tuple(1, 2.3, "5"), Tuple(1, 2.3, "4"))
+    XCTAssertNotEqual(Tuple(1, 2.9, "4"), Tuple(1, 2.3, "4"))
+    XCTAssertNotEqual(Tuple(0, 2.3, "4"), Tuple(1, 2.3, "4"))
+    // This effectively tests Hashable too; we know the conformance synthesizer
```
Comment (saeta) [inline Tests/PenguinStructuresTests/TupleTests.swift:100]: I'll take your word on this one...

File: Sources/PenguinStructures/Tuple.swift [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    return lhs.tail < rhs.tail
+  }
+}
+
+private let prefixLength = "Tuple(".count
+
+extension Tuple: CustomStringConvertible {
+  public var description: String {
+    if Tail.self == Empty.self {
+      return "Tuple(\(String(reflecting:head )))"
```
Comment (saeta) [inline Sources/PenguinStructures/Tuple.swift:93]:
```suggestion
      return "Tuple(\(String(reflecting: head)))"
```

File: Tests/PenguinStructuresTests/TupleTests.swift (hunk 3 above)
Comment (dabrahams) [reply Tests/PenguinStructuresTests/TupleTests.swift:165]: Agh, thanks.  I knew I was going to forget that!

File: Sources/PenguinStructures/Tuple.swift (hunk 4 above)
Comment (dabrahams) [reply Sources/PenguinStructures/Tuple.swift:104]: Yep, good plan. #40

Comment (dabrahams) [reply Sources/PenguinStructures/Tuple.swift:104]: #40

File: Tests/PenguinStructuresTests/TupleTests.swift (hunk 5 above)
Comment (dabrahams) [reply Tests/PenguinStructuresTests/TupleTests.swift:100]: By testing comparison we've verified that synthesized conformance is considering all the values the tuple is initialized with. Doing more would amount to testing the compiler, which is not what we need to do here.

**Pull request merged**